swift: support Linux (test build)

Bo98 opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

This will be upstreamed in https://github.com/Homebrew/homebrew-core/pull/73984, but I want to steal your CI first to test out if anything I did actually worked.

danielnachun commented:
I spent a long time working on this myself and eventually figured out that Swift requires Python 2 until the 5.4 release.  It's not used for building but instead for the GYB macro system.  The macros have been ported to Python 3 in the development branch of Swift 5.4, but those changes won't land in a stable release until 5.4 is released.  We can try to get this PR to work up to the point where it fails for Python 2, but I'm fairly sure that's as far as we'll get.

To fix the current error, I think you need to add `include Language::Python::Virtualenv` so you can install the Python build deps.

Bo98 commented:
> I spent a long time working on this myself and eventually figured out that Swift requires Python 2 until the 5.4 release.

I suspected there might be something that would block this - I just didn't know where it was at yet since this was my first attempt at it. I'll maybe keep some of the code here for the future if it's useful enough - depends on what others think.

> I think you need to add `include Language::Python::Virtualenv` so you can install the Python build deps.

Yeah, this is just me not testing things properly - I'll fix that.

Bo98 commented:
I actually expect Swift 5.4 to be released very soon (included in Xcode 12.5 which will release imminently), so won't be long to wait.

danielnachun commented:
> I actually expect Swift 5.4 to be released very soon (included in Xcode 12.5 which will release imminently), so won't be long to wait.

That's great news!  In that case we should be close to getting this supported.  

On a related note, we'll have our work cut out for us to figure out which packages that depend on Swift can be built without Xcode.  Some may have dependencies on macOS frameworks and will never work on Linux, but others may work with Linux build tools.  I'm nearly finished testing all the unbottled formulae on Linux so I'll have a list of these packages very soon.

Bo98 commented:
> Some may have dependencies on macOS frameworks and will never work on Linux

I imagine quite a lot fall into this category. If you are intending to develop a cross-platform tool, you're probably not going to be using Swift.

But I do have the the ports of Foundation.framework (which also contains some of CoreFoundation) and libdispatch included here so you never know.

---

Anyway, I've switched this to try the 5.4 snapshots to see how well it works - no point trying to do anything on 5.3.3 if it's going to be old in a few days.

Bo98 commented:
What is the correct value that should be inserted here:

https://github.com/apple/swift/blob/c5b7e7fb9c2fe0694200015db87395617ae26633/stdlib/public/Platform/glibc.modulemap.gyb#L48

considering that we could be using brewed glibc or sytem glibc?

Bo98 commented:
It's just a UUID header so it probably doesn't matter I suppose.

danielnachun commented:
I'm not actually sure why they are referencing `glibc` there because `uuid/uuid.h` is provided by `util-linux`, not `glibc`.  That's also true on Arch Linux.  So we can just replace `GLIBC_INCLUDE_PATH` with `#{Formula["util-linux"].opt_include}` and we'd be covered regardless of which `glibc` we are using.

danielnachun commented:
The current issue seems to be the Swift's copy of LLVM it uses to build can't find the `icu4c` headers: 

```
fatal error: 'unicode/ustring.h' file not found
```

If we can figure out how to pass through we'll be one step closer!

Bo98 commented:
Ah the 5.4 branch has more under "glibc": https://github.com/apple/swift/blob/release/5.4/stdlib/public/Platform/glibc.modulemap.gyb. If it's going to be bits from multiple formulae, it's probably not worth touching from the default `/usr/include`.

danielnachun commented:
> Ah the 5.4 branch has more under "glibc": https://github.com/apple/swift/blob/release/5.4/stdlib/public/Platform/glibc.modulemap.gyb. If it's going to be bits from multiple formulae, it's probably not worth touching from the default `/usr/include`.

That does look tricky so we can leave it alone for now.  I will check the headers individually later but at a first glance, UUID seems to be the only one which is not provided by `glibc`.  Assuming that fixing this actually matters, the solution might end up being setting `GLIBC_INCLUDE_PATH` to the brewed `glibc` if the user has that installed, and then for any headers not actually provided by `glibc`, replacing `GLIBC_INCLUDE_PATH` with the `opt_include` for the formula which does provide it.  

The way they are actually implementing this is not ideal as it assumes a single shared include dir, so if this causes us grief I will consider filing an upstream bug asking them to use separate variables for headers not actually provided by `glibc`.

danielnachun commented:
Now this is failing because it can't find the `gcc` runtime libs:

```
ld.gold: error: cannot find -lstdc++
ld.gold: error: cannot find -lgcc_s
ld.gold: error: cannot find -lgcc
ld.gold: error: cannot find -lgcc_s
ld.gold: error: cannot find -lgcc
```

In CI these are actually provided by the system at `/usr/lib/x86_64-linux-gnu`.  This might cause issues if someone has to build from source locally, but let's see if we can at least get it to find the libraries in CI first.

Bo98 commented:
I'm somewhat surprised `ld.gold` doesn't search that by default.

danielnachun commented:
Yeah it's weird that `ld.gold` wouldn't search there.  The directory handling of the GCC runtime libs is a fragmented mess though.  On Debian systems, it seems to be under `/usr/lib/PLATFORM_STRING`, while on RedHat (at least on the CentOS systems I use) they are directly under `/usr/lib64`, and on Arch Linux it's just in `/usr/lib`.  

I see you tried adding `gcc` as a dependency - hopefully it has an easier time finding the runtimes libs. Alternatively, we could see if it's possible to use `ld` instead.  I'm not sure if `ld.gold` checks `/etc/ld.conf` so that may be part of the reason why it can't find the GCC runtime libs when using the system GCC.

Bo98 commented:
I avoided the problem (I hope) by passing `--swift-include-tests=0` since it looked like the bad linker call was in a test.

The GCC dep was in response to [these compile errors](https://github.com/Homebrew/linuxbrew-core/runs/2209850076?check_suite_focus=true#step:7:8948), since those errors look like it is simply incompatible with GCC 5.

Bo98 commented:
For posterity:

<details><summary>GCC 5 errors</summary>

```
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:73:18: error: could not convert ‘(const char*)"Builtin.Int"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int"};
                  ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:76:19: error: could not convert ‘(const char*)"Builtin.Int8"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int8"};
                   ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:79:20: error: could not convert ‘(const char*)"Builtin.Int16"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int16"};
                    ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:82:20: error: could not convert ‘(const char*)"Builtin.Int32"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int32"};
                    ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:85:20: error: could not convert ‘(const char*)"Builtin.Int64"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int64"};
                    ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:88:21: error: could not convert ‘(const char*)"Builtin.Int128"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int128"};
                     ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:91:21: error: could not convert ‘(const char*)"Builtin.Int256"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int256"};
                     ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:94:21: error: could not convert ‘(const char*)"Builtin.Int512"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Int512"};
                     ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:97:25: error: could not convert ‘(const char*)"Builtin.IntLiteral"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.IntLiteral"};
                         ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:100:21: error: could not convert ‘(const char*)"Builtin.FPIEEE"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.FPIEEE"};
                     ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:103:20: error: could not convert ‘(const char*)"Builtin.FPPPC"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.FPPPC"};
                    ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:106:27: error: could not convert ‘(const char*)"Builtin.NativeObject"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.NativeObject"};
                           ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:109:27: error: could not convert ‘(const char*)"Builtin.BridgeObject"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.BridgeObject"};
                           ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:112:25: error: could not convert ‘(const char*)"Builtin.RawPointer"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.RawPointer"};
                         ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:115:36: error: could not convert ‘(const char*)"Builtin.RawUnsafeContinuation"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.RawUnsafeContinuation"};
                                    ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:118:33: error: could not convert ‘(const char*)"Builtin.UnsafeValueBuffer"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     {"Builtin.UnsafeValueBuffer"};
                                 ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:121:18: error: could not convert ‘(const char*)"Builtin.Job"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Job"};
                  ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:127:28: error: could not convert ‘(const char*)"Builtin.UnknownObject"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.UnknownObject"};
                            ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:130:18: error: could not convert ‘(const char*)"Builtin.Vec"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Vec"};
                  ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:133:23: error: could not convert ‘(const char*)"Builtin.SILToken"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.SILToken"};
                       ^
/tmp/swift-20210327-10904-2vpa1i/swift-swift-5.4-DEVELOPMENT-SNAPSHOT-2021-03-25-a/include/swift/Strings.h:136:19: error: could not convert ‘(const char*)"Builtin.Word"’ from ‘const char*’ to ‘const llvm::StringLiteral’
     "Builtin.Word"};
                   ^
/tmp/swift-20210327-10904-2vpa1i/swift/lib/Demangling/Demangler.cpp: In member function ‘swift::Demangle::Node* swift::Demangle::Demangler::demangleWitness()’:
/tmp/swift-20210327-10904-2vpa1i/swift/lib/Demangling/Demangler.cpp:2743:32: error: expected primary-expression before ‘unsigned’
         case 'd': Directness = unsigned(Directness::Direct); break;
                                ^
/tmp/swift-20210327-10904-2vpa1i/swift/lib/Demangling/Demangler.cpp:2744:32: error: expected primary-expression before ‘unsigned’
         case 'i': Directness = unsigned(Directness::Indirect); break;
                                ^
```
</details>

Bo98 commented:
Ah the linker failures are back:

```
ld.gold: error: cannot find -lpthread
ld.gold: error: cannot find -ldl
ld.gold: error: cannot find -lstdc++
ld.gold: error: cannot find -lm
ld.gold: error: cannot find -lgcc_s
ld.gold: error: cannot find -lgcc
ld.gold: error: cannot find -lc
ld.gold: error: cannot find -lgcc_s
ld.gold: error: cannot find -lgcc
```

danielnachun commented:
You might have to try `-DSWIFT_USE_LINKER=lld` to force it to use `lld` if adding it as an `llvm_component` isn't enough. And if that fails `-DSWIFT_USE_LINKER=ld` might be able to force it to use the BFD linker.  

There are some comments in https://github.com/apple/swift/blob/2083e1bba1fa4245a08a800f1fc78a619da200ac/lib/Driver/UnixToolChains.cpp claiming that the BFD linker doesn't work properly... on the IBM s390 and some embedded ARM subset I've never heard of called Thumb.   So I don't think it would be problem for us!

Bo98 commented:
Yeah, I might have to use that. I do wonder though: have you have issues with `ld.gold` elsewhere? Is `ld.gold` broken?

danielnachun commented:
I'm not sure we even link anything with `ld.gold` in any formulae, right now, so it may very well be untested.  These old bug reports may be relevant: https://bugs.launchpad.net/ubuntu/+source/binutils/+bug/780455, https://bugs.launchpad.net/ubuntu/+source/binutils/+bug/738098.  It's possible that `ENV.append "LIBRARY_PATH", "/usr/lib/x86_64-linux-gnu"` could fix this with `ld.gold`.

Bo98 commented:
It might not necessarily be a superenv bug, but I do suspect the `tool == "ld"` [here](https://github.com/Homebrew/brew/blob/3ff5795ae105ffc09119ee897d828a1d20c110b9/Library/Homebrew/shims/super/cc#L108) should be `mode == :ld` instead.

The logs I've enabled in the second commit will give more info on what actually gets passed to `ld.gold`.

danielnachun commented:
You're probably right - we need to do some work to make the superenv work better with `clang` for compilation and and `ld.gold` and `lld` for linking on Linux, as those tools become more popular.  Of course this task is made harder by the fact that every big project that uses LLVM seems to think they should just fork it instead of trying to get their patches merged...

Bo98 commented:
`lld` isn't even shimmed at all in the superenv.

Bo98 commented:
https://github.com/Homebrew/brew/pull/10959#issuecomment-809021687

danielnachun commented:
Can we try removing the `gcc` dependency (and `fails_with`) after we merge https://github.com/Homebrew/brew/pull/10959?  I think the other issue was related to needing `-fpermissive` as a GCC arg, but I don't think that was specific to `gcc-10` and it may work with fine with the host `gcc-5`.

carlocab commented:
> every big project that uses LLVM seems to think they should just fork it instead of trying to get their patches merged...

💯

Bo98 commented:
Swift's better than most as they submit all their patches upstream except some LLDB stuff. The main problem is that the versions of LLVM they build their tools around do not align with LLVM stable releases. So Swift 5.4 currently uses Clang 10 with some Clang 11 patches.

>  I think the other issue was related to needing `-fpermissive` as a GCC arg

The GCC 5 errors didn't seem to mention anything about `-fpermissive` like the GCC 10 ones did.

danielnachun commented:
> Swift's better than most as they submit all their patches upstream except some LLDB stuff. The main problem is that the versions of LLVM they build their tools around do not align with LLVM stable releases. So Swift 5.4 currently uses Clang 10 with some Clang 11 patches.

I guess that's better though still not ideal.  Perhaps the other issue is that LLVM is less conservative about backwards incompatible changes with new releases than GCC, which I can imagine would create headaches for downstream projects.  

> > I think the other issue was related to needing `-fpermissive` as a GCC arg
> 
> The GCC 5 errors didn't seem to mention anything about `-fpermissive` like the GCC 10 ones did.

In that case maybe it's better to stick with GCC 10 until we can at least get it to build now that we've hopefully fixed the superenv handling of `ld.gold`.

Bo98 commented:
Hmm, interesting one:

```
swift/lib/Basic/Fingerprint.cpp:41:68: error: incomplete type '{anonymous}::SmallStringBound<llvm::SmallString<32> >' used in nested name specifier
   41 | static_assert(SmallStringBound<std::result_of<MD5Digest_t>::type>::value ==
      |                                                                    ^~~~~
```

Tried to reduce this to a small test case that fails on GCC but I haven't been able to yet.

danielnachun commented:
> Hmm, interesting one:
> 
> ```
> swift/lib/Basic/Fingerprint.cpp:41:68: error: incomplete type '{anonymous}::SmallStringBound<llvm::SmallString<32> >' used in nested name specifier
>    41 | static_assert(SmallStringBound<std::result_of<MD5Digest_t>::type>::value ==
>       |                                                                    ^~~~~
> ```
> 
> Tried to reduce this to a small test case that fails on GCC but I haven't been able to yet.
> 
> I see that this code has been rewritten on `master` so I could potentially patch it with that change.

Haha I guess this is why we usually wait until the stable release.  Since we have no choice right now but to use the preview release and this is a draft, I'm in favor of doing "whatever it takes" to get a working build, so long as we know what to revisit once the stable release is out.

Bo98 commented:
To be clear `master` means Swift 5.5 now - not 5.4.

5.4 code is pretty much frozen at this point. There's not been any 5.4 changes since the snapshot I'm using.

Bo98 commented:
Hmm, I'm thinking this simply doesn't compile on GCC and requires us to use Clang instead.

danielnachun commented:
Could it be that we need to set `ENV.O0`: https://stackoverflow.com/questions/54959058/when-we-mustnt-declare-function-inlinegcc-compiler?  That should disable inlining, which might overcome that error.

If that doesn't work I think we'll have to figure out how to use LLVM instead.  FWIW on Arch Linux they do use Swift for their unofficial package: https://aur.archlinux.org/packages/swift-language/.

Bo98 commented:
Swift is using `__attribute__((always_inline))` in this case, so optimisation level won't cover it. It's a hard error.

danielnachun commented:
...and we hit the 6 hour time limit.  At the very least that tells us you were right to switch to `clang` for building!  I guess this goes in the list of the formula that need to be built with the self-hosted VM.  I don't know how to use the self-hosted VMs yet but we can summon @iMichka for this.

iMichka commented:
In a few days maybe, I'm still working on the deployment scripts.

danielnachun commented:
In that case @Bo98 if you want to test locally use `docker run -it homebrew/ubuntu16.04:master` to get our CI environment in a Docker image.  I can also check out your branch and test locally on my machine if that's easier.

Bo98 commented:
It might be a few days before I can afford the time to let it run locally.

danielnachun commented:
I checked out your branch and ran it locally yesterday - here are the gist-logs: https://gist.github.com/danielnachun/91c5c5a0e28487588c960a9d333b95a0.

Using `clang` brought us **much** further into the build process.  The failure we hit now is actually happening when using `swiftc` to build `libdispatch`: 

```
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:1:8: error: redefinition of module 'Dispatch'
module Dispatch {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:1:8: note: previously defined here
module Dispatch {
       ^
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:8:8: error: redefinition of module 'DispatchIntrospection'
module DispatchIntrospection [system] [extern_c] {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:8:8: note: previously defined here
module DispatchIntrospection [system] [extern_c] {
       ^
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:13:8: error: redefinition of module 'CDispatch'
module CDispatch [system] [extern_c] {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:13:8: note: previously defined here
module CDispatch [system] [extern_c] {
       ^
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:1:8: error: redefinition of module 'Dispatch'
module Dispatch {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:1:8: note: previously defined here
module Dispatch {
       ^
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:8:8: error: redefinition of module 'DispatchIntrospection'
module DispatchIntrospection [system] [extern_c] {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:8:8: note: previously defined here
module DispatchIntrospection [system] [extern_c] {
       ^
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/dispatch/module.modulemap:13:8: error: redefinition of module 'CDispatch'
module CDispatch [system] [extern_c] {
       ^
/tmp/swift-20210330-22695-18jv60z/swift-corelibs-libdispatch/dispatch/module.modulemap:13:8: note: previously defined here
module CDispatch [system] [extern_c] {
       ^
<unknown>:0: error: could not build C module '_SwiftDispatchOverlayShims'
ninja: build stopped: subcommand failed.
```

The best I can find for this error from Google is that it can happen if one ends up including two header files with conflicting definitions, but that may be irrelevant.  Any thoughts?  I'm happy to serve as "human CI" for a bit just to see if we can get the built to finish!

Bo98 commented:
Thanks for testing it out! Seems like it has already built and installed libdispatch but gets confused which files to use (local build or installed build) while building `yams`. I'll poke around and see how the build system works there.

Bo98 commented:
Latest push should fix it. Diff: https://github.com/Homebrew/linuxbrew-core/compare/bb7812b0f9b6a46df1bcc424a4d37b02e99b81d7..8078591ac9dc0269bccb2d80603e4c88cf25c61e

(Removed the `-fpermissive` patch too since we won't need that when using Clang.)

danielnachun commented:
Could you squash that last commit into the first so we just have a single commit?

danielnachun commented:
Oh whoops I misunderstood stood that.  I'm reseting my Docker VM and starting the build again.  Let's see how far it goes!

danielnachun commented:
We're on the home stretch now.  I think Swift itself builds now, but there's a problem with another module finding the `sqlite` headers.

https://gist.github.com/c4e1e18a128c781555253d0a89139985

```
[103/111] Compiling TSCUtility Archiver.swift
<module-includes>:1:10: note: in file included from <module-includes>:1:
#include "csqlite3.h"
         ^
/tmp/swift-20210330-15849-10jzh96/swift-tools-support-core/Sources/TSCclibc/include/csqlite3.h:1:10: error: 'sqlite3.h' file not found
#include <sqlite3.h>
         ^
/tmp/swift-20210330-15849-10jzh96/swift-tools-support-core/Sources/TSCUtility/SQLite.swift:14:29: error: could not build C module 'CSQLite3'
@_implementationOnly import CSQLite3
```

Maybe we need more `EXTRA_CMAKE_ARGS`?

Bo98 commented:
Hmm, a tricky one since:

> `-- Found SQLite3: /home/linuxbrew/.linuxbrew/include (found version "3.35.2")`

It seems to find it without issue when used with a C/C++ compiler, but there's issues when under the Swift compiler.

I wonder if it would be a terrible idea to insert into the default LLVM include path. We currently don't, even in the standard LLVM formulae.

carlocab commented:
> I wonder if it would be a terrible idea to insert `HOMEBREW_PREFIX/include` into the default LLVM include path.

I've been thinking about doing this. It would make the experience of using the `llvm` formula more consistent between ARM and Intel on macOS.

Bo98 commented:
> It would make the experience of using the `llvm` formula more consistent between ARM and Intel on macOS.

On macOS, even `/usr/local` isn't in the default search path:

```
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/opt/llvm/bin/../include/c++/v1
 /usr/local/Cellar/llvm/11.1.0/lib/clang/11.1.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

Bo98 commented:
Alternative fix for the less brave: just patch the Csqlite3 modulemap to not use their weird `csqlite3.h` and instead point directly to the sqlite3.h header with an absolute path.

Bo98 commented:
For example, what I've just pushed.

(Diff: https://github.com/Homebrew/linuxbrew-core/compare/a16949daee0b03964fea202a9d908efd2b830c0b..86f9e29634cf3399d51aa1ca60a00223181ce1ec)

carlocab commented:
> On macOS, even `/usr/local` isn't in the default search path:

Oops. I had forgotten I had set `CPPFLAGS`, etc in my startup files.

Bo98 commented:
CMake has Swift support it seems, so CMake should be handling adding the correct include flags. Not really sure what's happening under the hood.

danielnachun commented:
> CMake has Swift support it seems, so CMake should be handling adding the correct include flags. Not really sure what's happening under the hood.

Is there anything I can pull out of the build dir?  I ran it with `--debug`.  Maybe a `CMakeLists`?

carlocab commented:
You don't have a CMake dependency here, though. Does Swift vendor its own copy of CMake?

Bo98 commented:
It's a few lines above the Xcode dep

danielnachun commented:
I'll try to new change to see if we can bypass that weird `csqlite3.h` issue in that module map.  Luckily I can leave this running in the background building while I do other things.

Bo98 commented:
Alternatively, we can try this patch I just put together: https://github.com/Bo98/swift-tools-support-core/commit/33ba31c46bf6ad7b703a0cbbcb905854f6733428

Bo98 commented:
Sorry for switching at the last minute. It's worth trying that patch instead as it's upstreamable as is.

danielnachun commented:
I do see in https://github.com/apple/swift/blob/bce662067adab5d4bb6083add7b91567a0f86415/docs/WindowsBuild.md that they pass `SQLite3_INCLUDE_DIR` and `SQLite3_LIBRARY`, so it makes sense that we'd probably have to do something with those variables.

Bo98 commented:
That should just help CMake find it, and CMake already does automatically. It's just not passed properly to the Swift compiler, which is an error in the CMake file. But since it looks like the modulemap is installed to disk, us patching that modulemap might end up help Swift find SQLite3 at runtime too.

danielnachun commented:
That patch did indeed work, and now we have an issue with the linker finding what I assume is `icu4c`:

https://gist.github.com/6ae38d3c064b42a866869ffa69c5b97b

```
[130/135] Compiling swift_run main.swift
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_charAge_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_charName_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_getIntPropertyValue_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_getNumericValue_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_hasBinaryProperty_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_strToLower_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_strToTitle_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'u_strToUpper_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_close_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_following_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_open_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_preceding_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_setText_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'ubrk_setUText_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'unorm2_getNFCInstance_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'unorm2_hasBoundaryBefore_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'unorm2_normalize_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'unorm2_spanQuickCheckYes_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'utext_openUChars_68'
/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux/libswiftCore.so: error: undefined reference to 'utext_openUTF8_68'
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)
[130/135] Linking swift-package-collection
[130/135] Linking swift-package
[130/135] Linking swift-test
[130/135] Linking swift-run
[130/135] Linking swift-build
env SWIFTCI_USE_LOCAL_DEPS=1 SWIFTPM_MACOS_DEPLOYMENT_TARGET=10.15 LD_LIBRARY_PATH=/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/tsc/lib:/tmp/swift-20210331-18439-1kwd4p1/build/llbuild-linux-x86_64/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/yams/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/swift-argument-parser/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/swift-driver/lib:/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux SWIFT_EXEC=/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/bin/swiftc SWIFT_DRIVER_SWIFT_EXEC=/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/bin/swiftc SWIFTPM_PD_LIBS=/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/pm /tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/bin/swift-build --disable-sandbox --build-path /tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64 --configuration release -Xlinker -rpath -Xlinker $ORIGIN/../lib/swift/linux -Xswiftc -module-cache-path -Xswiftc /tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/module-cache -Xmanifest -module-cache-path -Xmanifest /tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/module-cache
--- bootstrap: error: Command '['env', 'SWIFTCI_USE_LOCAL_DEPS=1', 'SWIFTPM_MACOS_DEPLOYMENT_TARGET=10.15', 'LD_LIBRARY_PATH=/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/tsc/lib:/tmp/swift-20210331-18439-1kwd4p1/build/llbuild-linux-x86_64/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/yams/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/swift-argument-parser/lib:/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/swift-driver/lib:/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/lib/swift/linux', 'SWIFT_EXEC=/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/bin/swiftc', 'SWIFT_DRIVER_SWIFT_EXEC=/home/linuxbrew/.linuxbrew/Cellar/swift/5.4/libexec/bin/swiftc', 'SWIFTPM_PD_LIBS=/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/pm', '/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/bootstrap/bin/swift-build', '--disable-sandbox', '--build-path', '/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64', '--configuration', 'release', '-Xlinker', '-rpath', '-Xlinker', '$ORIGIN/../lib/swift/linux', '-Xswiftc', '-module-cache-path', '-Xswiftc', '/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/module-cache', '-Xmanifest', '-module-cache-path', '-Xmanifest', '/tmp/swift-20210331-18439-1kwd4p1/build/swiftpm-linux-x86_64/module-cache']' returned non-zero exit status 1.
Building the standard library for: swift-stdlib-linux-x86_64
```

Bo98 commented:
That'll be a needle in a haystack - it could stem back from an issue quite far back in the build.

Bo98 commented:
Will probably not fix anything but we should depend on `pkg-config` rather than pass in the paths like we are now.

Bo98 commented:
I've simplified the formula a bit (including the `fails_with` now that Homebrew/brew#10968 is merged) and added a temporary hack to enable verbose building of SwiftPM.

danielnachun commented:
Okay I'll restart the test and see if the logs are more informative this time.  I didn't post this yesterday but I did a quick check of the `libSwiftCore.so` binary in the temp dir and I did see `$HOMEBREW_PREFIX/opt/icu4c/lib` in the `rpath` with `patchelf`.  But I'm not sure if that was actually the `libSwiftCore.so` that is failing in the install, do may not really have been that informative.

danielnachun commented:
I think something got lost in the simplification because now it fails to find the `icu4c` headers:

https://gist.github.com/d514d2d35c56a71a02702e39d6c9d7a8

```
/tmp/swift-20210331-15172-v9u524/swift/stdlib/public/stubs/UnicodeNormalization.cpp:78:10: fatal error: 'unicode/ustring.h' file not found
#include <unicode/ustring.h>
         ^~~~~~~~~~~~~~~~~~~
```

I guess we still need those extra CMake args?

Bo98 commented:
Hmm, it seems like it doesn't cope with `HOMEBREW_PREFIX/include` very well, but seems to have no problem with `opt_include`. Not entirely sure why.

I have spotted an error in my previous CMake options however. I've fixed that now so it should hopefully fix linkage with `icu4c`.

danielnachun commented:
Here are the latest build logs: https://gist.github.com/danielnachun/7502cd2bfc7b67ad932501e5b271106a.

Now it can't find `ncurses`:

```
/home/linuxbrew/.linuxbrew/bin/ld: cannot find -lcurses
```

Unfortunately the log formatting is kind of messed up so you'll have to use some regex replacement to make it readable.

Bo98 commented:
> `/home/linuxbrew/.linuxbrew/bin/clang++`

That's a bit weird. `/home/linuxbrew/.linuxbrew/bin` isn't even in the superenv PATH...

I've pushed something that'll probably fix it.

Bo98 commented:
Out of interest, what situation is self-hosted Linux CI in? Swift 5.4 release is imminent (as in, it could be today even) so I'm wondering what to upstream here to homebrew-core.

iMichka commented:
Right now I am working on this: https://github.com/Homebrew/homebrew-core/pull/75949
Once this is done I'll be able to add the runner to the homebrew-core repo. It might take a few more days/before before we have something up and running. I am unsure we will add any runner to the linuxbrew-core repo (if anyone wants to do it, you can pick it up, but I prefer to focus on working on the core merge first).

Bo98 commented:
I broke the branch (fixed now but can't re-open).

I'll find time to test this on Linux at some point. I've upstreamed my changes to the Homebrew/core PR but maybe I should just drop it.

danielnachun commented:
I definitely want to get this figured out, but I think it will be a lot easier once we have the self-hosted Linux runner in homebrew/core.
